src/lib/ams.c

```diff
@@ -0,0 +1,749 @@
+#include <sof/drivers/idc.h>
+#include <sof/lib/alloc.h>
+#include <sof/lib/memory.h>
+#include <errno.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <sof/lib/ams.h>
+
+LOG_MODULE_REGISTER(ams, CONFIG_SOF_LOG_LEVEL);
+
+DECLARE_SOF_UUID("ams", ams_uuid, 0xea9c4bca, 0x5b7d, 0x48c6,
+		 0x95, 0x86, 0x55, 0x3e, 0x27, 0x23, 0x5b, 0xeb);
+
+DECLARE_TR_CTX(ams_tr, SOF_UUID(ams_uuid), LOG_LEVEL_INFO);
+
+/* AMS routing table */
+const uint8_t g_ams_core_routing[CONFIG_CORE_COUNT][CONFIG_CORE_COUNT] = {
+/* to core: 0..N */
+/* from core 1 */ {
+	0,
+	1,
+	2,
+	}, /* master core can target every other core directly */
+/* from core 2 */ {
+	0,
+	1,
+	0,
+	}, /* and slaves must always go thru master core */
+/* from core 3 */ {
+	0,
+	0,
+	2,
+}};
+
+static struct am_service_context ctx;
+static struct am_service_shared_context shared_ctx;
+static struct ams_routing_entry rt_table[AM_ROUTING_TAB_SIZE];
+static struct ams_producer producer_table[AM_ROUTING_TAB_SIZE];
+static struct uuid_idx uuid_table[AM_SERVICE_UUID_TABLE_SIZE];
+
+void ams_acquire(struct am_service_context *ctx)
+{
+	struct coherent __sparse_cache *c = coherent_acquire_thread(&ctx->shared->c,
+								    sizeof(ctx->shared));
+
+	ctx->shared = attr_container_of(c, struct am_service_shared_context __sparse_cache,
+					c, __sparse_cache);
+	dcache_invalidate_region(ctx->shared->routing_table,
+				 ctx->shared->routing_table_size * sizeof(*ctx->shared->routing_table));
+	dcache_invalidate_region(ctx->shared->uuid_table,
+				 ctx->shared->uuid_table_size * sizeof(ctx->shared->uuid_table));
+	dcache_invalidate_region(ctx->shared->producer_table,
+				 ctx->shared->producer_table_size * sizeof(ctx->shared->producer_table));
+}
+
+void ams_release(struct am_service_context *ctx)
+{
+	dcache_writeback_invalidate_region(ctx->shared->producer_table,
+					   ctx->shared->producer_table_size * sizeof(ctx->shared->producer_table));
+	dcache_writeback_invalidate_region(ctx->shared->uuid_table,
+					   ctx->shared->uuid_table_size * sizeof(ctx->shared->uuid_table));
+	dcache_writeback_invalidate_region(ctx->shared->routing_table,
+					   ctx->shared->routing_table_size * sizeof(*ctx->shared->routing_table));
+	coherent_release_thread(&ctx->shared->c, sizeof(ctx->shared));
+}
+
+static int am_service_message_send_over_ixc(struct async_message_service *ams, uint32_t slot,
+					    struct ams_routing_entry *target,
+					    void *ctx)
+{
+#if CONFIG_SMP
+	send_message_over_ixc(ams, slot, target);
+#else
+	return -EINVAL;
+#endif
+	return 0;
+}
+
+const struct uuid_idx *am_service_find_uuid_entry_by_uuid(struct am_service_context *const am_service_ctx,
+							  uint8_t const uuid[AM_MESSAGE_UUID_SIZE])
+{
+	uint32_t index;
+
+	if (uuid == NULL)
+		return NULL;
+
+	/* try to find existing entry */
+	for (index = 0; index < am_service_ctx->shared->uuid_table_size; index++) {
+		if (memcmp(am_service_ctx->shared->uuid_table[index].message_uuid,
```

Comment:
the compiler will *probably* optimise this, but to help the reader it would be good to use a local variable to point to `am_service_ctx->shared->uuid_table`